File: c++/583. Delete Operation for Two Strings.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int n1 = word1.size(), n2 = word2.size();
        vector<vector<int>> dp(n1+1, vector<int>(n2+1, 0));
        for (int i = 1; i <= n1; i++) {
            for (int j = 1; j <=n2; j++) {
                if (word1[i-1] == word2[j-1])
                    dp[i][j] = dp[i-1][j-1] + 1;
                else
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
            }
        }

        int lcs = dp[n1][n2];
        return n1 + n2 - 2*lcs;
    }
};

// leetcode. 1143 求最长公共子序列

File: go/src/typical_problems/DP/583. Delete Operation for Two Strings/583. Delete Operation for Two Strings.go
package main

import (
	"fmt"
	"math"
)

func minDistance(word1 string, word2 string) int {
	n1, n2 := len(word1), len(word2)
	f := make([][]int, n1 + 1)
	for i := range f {
		f[i] = make([]int, n2 + 1)
		for j := range f[i] {
			f[i][j] = math.MaxInt32 >> 1
		}
	}
	f[0][0] = 0
	for i := 1; i <= n1; i++ {
		f[i][0] = i
	}
	for j := 1; j <= n2; j++ {
		f[0][j] = j
	}
	for i := 1; i <= n1; i++ {
		for j := 1; j <= n2; j++ {
			if word1[i-1] == word2[j-1] {
				f[i][j] = min(f[i][j], f[i-1][j-1])
			} else {
				f[i][j] = min(f[i-1][j], f[i][j-1]) + 1
			}
		}
	}

	return f[n1][n2]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	fmt.Println(minDistance("sea", "eat"))
}
